docs/using/runconfig.md
# Run Configuration

This page gives an overview of the configuration options available for a pipeline run.

## Default Configuration File

Below is an example of a default `config.yaml` file. Note that no images or other input files have been provided. The file can be either edited directly or through the editor available on the run detail page.

!!! warning
    Similarly to Python files, the indentation in the run configuration YAML file is important as it defines nested parameters.

!!! example "config.yaml"
    ```yaml
    # This file specifies the pipeline configuration for the current pipeline run.
    # You should review these settings before processing any images - some of the default
    # values will probably not be appropriate.

    run:
      # Path of the pipeline run
      path: ... # auto-filled by pipeline initpiperun command

      # Hide astropy warnings during the run execution.
      suppress_astropy_warnings: True

    inputs:
      # NOTE: all the inputs must match with each other, i.e. the catalogue for the first
      # input image (inputs.image[0]) must be the first input catalogue (inputs.selavy[0])
      # and so on.
      image:
      # list input images here, e.g. (note the leading hyphens)
      # - /path/to/image1.fits
      # - /path/to/image2.fits

      selavy:
      # list input selavy catalogues here, as above with the images

      noise:
      # list input noise (rms) images here, as above with the images

      # Required only if source_monitoring.monitor is true, otherwise optional. If not providing
      # background images, remove the entire background section below.
      background:
      # list input background images here, as above with the images


    source_monitoring:
      # Source monitoring can be done both forward and backward in 'time'.
      # Monitoring backward means re-opening files that were previously processed and can be slow.
      monitor: True

      # Minimum SNR ratio a source has to be if it was placed in the area of minimum rms in
      # the image from which it is to be extracted from. If lower than this value it is skipped
      min_sigma: 3.0
  
      # Multiplicative scaling factor to the buffer size of the forced photometry from the
      # image edge
      edge_buffer_scale: 1.2

      # Passed to forced-phot as `cluster_threshold`. See docs for details. If unsure, leave
      # as default.
      cluster_threshold: 3.0

      # Attempt forced-phot fit even if there are NaN's present in the rms or background maps.
      allow_nan: False

    source_association:
      # basic, advanced, or deruiter
      method: basic

      # Maximum source separation allowed during basic and advanced association in arcsec
      radius: 10.0

      # Options that apply only to deruiter association
      deruiter_radius: 5.68  # unitless
      deruiter_beamwidth_limit: 1.5  # multiplicative factor

      # Split input images into sky region groups and run the association on these groups in
      # parallel. Best used when there are a large number of input images with multiple
      # non-overlapping patches of the sky.
      # Not recommended for smaller searches of <= 3 sky regions.
      parallel: False

      # If images have been submitted in epoch dictionaries then an attempt will be made by
      # the pipeline to remove duplicate sources. To do this a crossmatch is made between
      # catalgoues to match 'the same' measurements from different catalogues. This
      # parameter governs the distance for which a match is made in arcsec. Default is 2.5
      # arcsec which is typically 1 pixel in ASKAP images.
      epoch_duplicate_radius: 2.5  # arcsec

    new_sources:
      # Controls when a source is labelled as a new source. The source in question must meet
      # the requirement of: min sigma > (source_peak_flux / lowest_previous_image_min_rms)
      min_sigma: 5.0

    measurements:
      # Source finder used to produce input catalogues. Only selavy is currently supported.
      source_finder: selavy

      # Minimum error to apply to all flux measurements. The actual value used will either
      # be the catalogued value or this value, whichever is greater. This is a fraction, e.g.
      # 0.05 = 5% error, 0 = no minimum error.
      flux_fractional_error: 0.0

      # Replace the selavy errors with Condon (1997) errors.
      condon_errors: True

      # Sometimes the local rms for a source is reported as 0 by selavy.
      # Choose a value to use for the local rms in these cases in mJy/beam.
      selavy_local_rms_fill_value: 0.2

      # Create 'measurements.arrow' and 'measurement_pairs.arrow' files at the end of 
      # a successful run.
      write_arrow_files: False

      # The positional uncertainty of a measurement is in reality the fitting errors and the
      # astrometric uncertainty of the image/survey/instrument combined in quadrature.
      # These two parameters are the astrometric uncertainty in RA/Dec and they may be different.
      ra_uncertainty: 1.0 # arcsec
      dec_uncertainty: 1.0  # arcsec

    variability:
      # For each source, calculate the measurement pair metrics (Vs and m) for each unique
      # combination of measurements.
      pair_metrics: True

      # Only measurement pairs where the Vs metric exceeds this value are selected for the
      # aggregate pair metrics that are stored in Source objects.
      source_aggregate_pair_metrics_min_abs_vs: 4.3

    processing:
      # Options to control use of Dask parallelism
      # NOTE: These are advanced options and you should only change them if you know what you are doing.

      # The total number of workers available to Dask ('null' means use one less than all cores)
      num_workers: null

      # The number of workers to use for disk IO operations (e.g. when reading images for forced extraction)
      num_workers_io: 5

      # The default maximum size (in MB) to allow per partition of Dask DataFrames
      # Increasing this will create fewer partitions and will potentially increase the memory footprint
      # of parallel tasks.
      max_partition_mb: 15

    ```

!!! note
    Throughout the documentation we use dot-notation to refer to nested parameters, for example `inputs.image` refers to the list of input images.
    This page on [YAML syntax](https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html) from the Ansible documentation is a good brief primer on the basics.

## Configuration Options

### General Run Options

**`run.path`**
Path to the directory for the pipeline run. This parameter will be automatically filled if the configuration file is generate with the [`initpiperun` management command](../adminusage/cli.md#initpiperun) or if the run was created with the web interface.

**`run.suppress_astropy_warnings`**
Boolean. Astropy warnings are suppressed in the logging output if set to `True`. Defaults to `True`.

### Input Images and Selavy Files

!!! warning "Warning: Entry Order"
    The order of the the inputs must be consistent between the different input types.
    I.e. if `image1.fits` is the first listed image then `image1_selavy.txt` must be the first selavy input listed.

**`inputs.image`**
Line entries or epoch headed entries. 
The full paths to the image FITS files to be processed - these can be regular FITS files, or FITS files that use a [`CompImageHDU`](https://docs.astropy.org/en/stable/io/fits/api/images.html#astropy.io.fits.CompImageHDU). In principle the pipeline also supports [`.fits.fz`](https://heasarc.gsfc.nasa.gov/fitsio/fpack/) files, although this is not officially supported. Epoch mode is activated by including an extra key value with the epoch name, see the example below for a demonstration.
Refer to [this section](../design/association.md#epoch-based-association) of the documentation for more information on epoch based association.

<!-- markdownlint-disable MD046 -->
!!! example "config.yaml"
    === "Normal mode"

        ```yaml
        inputs:
          image:
          - /full/path/to/image1.fits
          - /full/path/to/image2.fits
          - /full/path/to/image3.fits
        ```

    === "Epoch mode"

        ```yaml
        inputs:
          image:
            epoch01:
            - /full/path/to/image1.fits
            - /full/path/to/image2.fits
            epoch02:
            - /full/path/to/image3.fits
        ```
        Make sure that the epoch names are orderable in the correct order, for example, use:
        ```
        epoch01:
        ...
        epoch09:
        epoch10:
        ``` 
        and not:
        ```
        epoch1:
        ...
        epoch9:
        epoch10:
        ```
    <!-- markdownlint-enable MD046 -->

**`inputs.selavy`**
Line entries or epoch headed entries. 
The full paths to the selavy text files to be processed. 
Epoch mode is activated by including an extra key value with the epoch name, see the example below for a demonstration.
Refer to [this section](../design/association.md#epoch-based-association) of the documentation for more information on epoch based association.

!!! example "config.yaml"
    <!-- markdownlint-disable MD046 -->
    === "Normal mode"

        ```yaml
        inputs:
          selavy:
          - /full/path/to/image1_selavy.txt
          - /full/path/to/image2_selavy.txt
          - /full/path/to/image3_selavy.txt
        ```

    === "Epoch mode"

        ```yaml
        inputs:
          selavy:
            epoch01:
            - /full/path/to/image1_selavy.txt
            - /full/path/to/image2_selavy.txt
            epoch02:
            - /full/path/to/image3_selavy.txt
        ```
        Make sure that the epoch names are orderable in the correct order, for example, use:
        ```
        epoch01:
        ...
        epoch09:
        epoch10:
        ``` 
        and not:
        ```
        epoch1:
        ...
        epoch9:
        epoch10:
        ```
    <!-- markdownlint-enable MD046 -->

**`inputs.noise`**
Line entries or epoch headed entries. 
The full paths to the image noise (RMS) FITS files to be processed. 
Epoch mode is activated by including an extra key value with the epoch name, see the example below for a demonstration.
Refer to [this section](../design/association.md#epoch-based-association) of the documentation for more information on epoch based association.

!!! example "config.yaml"
    <!-- markdownlint-disable MD046 -->
    === "Normal mode"

        ```yaml
        inputs:
          noise:
          - /full/path/to/image1_rms.fits
          - /full/path/to/image2_rms.fits
          - /full/path/to/image3_rms.fits
        ```

    === "Epoch mode"

        ```yaml
        inputs:
          noise:
            epoch01:
            - /full/path/to/image1_rms.fits
            - /full/path/to/image2_rms.fits
            epoch02:
            - /full/path/to/image3_rms.fits
        ```
        Make sure that the epoch names are orderable in the correct order, for example, use:
        ```
        epoch01:
        ...
        epoch09:
        epoch10:
        ``` 
        and not:
        ```
        epoch1:
        ...
        epoch9:
        epoch10:
        ```
    <!-- markdownlint-enable MD046 -->

**`inputs.background`**
Line entries or epoch headed entries.
The full paths to the image background (mean) FITS files to be processed. 
Epoch mode is activated by including an extra key value with the epoch name, see the example below for a demonstration.
Refer to [this section](../design/association.md#epoch-based-association) of the documentation for more information on epoch based association.
The background images are only required to be defined if `source_monitoring.monitor` is set to `True`.

!!! example "config.yaml"
    <!-- markdownlint-disable MD046 -->
    === "Normal mode"

        ```yaml
        inputs:
          background:
          - /full/path/to/image1_bkg.fits
          - /full/path/to/image2_bkg.fits
          - /full/path/to/image3_bkg.fits
        ```

    === "Epoch mode"

        ```yaml
        inputs:
          background:
            epoch01:
            - /full/path/to/image1_bkg.fits
            - /full/path/to/image2_bkg.fits
            epoch02:
            - /full/path/to/image3_bkg.fits
        ```
        Make sure that the epoch names are orderable in the correct order, for example, use:
        ```
        epoch01:
        ...
        epoch09:
        epoch10:
        ``` 
        and not:
        ```
        epoch1:
        ...
        epoch9:
        epoch10:
        ```
    <!-- markdownlint-enable MD046 -->

#### Using glob expressions

Instead of providing each input file explicitly, the inputs can be given as glob expressions which are resolved and sorted. Glob expressions must be provided as a mapping with the key `glob`. Both normal and epoch mode are supported.

For example, the image input examples given above can be equivalently specified with the following glob expressions.

!!! example "config.yaml"
    <!-- markdownlint-disable MD046 -->
    === "Normal mode"

        ```yaml
        inputs:
          image:
            glob: /full/path/to/image*.fits
        ```

    === "Epoch mode"

        ```yaml
        inputs:
          image:
            epoch01:
              glob: /full/path/to/image[12].fits
            epoch02:
            - /full/path/to/image3.fits
        ```
    <!-- markdownlint-enable MD046 -->

Multiple glob expressions can also be provided as a list, in which case they are resolved and sorted in the order they are given. For example:

<!-- markdownlint-disable MD038 -->
!!! example "config.yaml"
    ```yaml
    inputs:
      image:
        glob:
        - /full/path/to/A/image*.fits
        - /full/path/to/B/image*.fits
    ```
<!-- markdownlint-enable MD038 -->

Note that it is not valid YAML to mix a sequence/list and a mapping/dictionary, meaning that for each input type (or epoch if using epoch mode), the files may be given _either_ as glob expressions or explicit file paths. For example, the following is **invalid**:

<!-- markdownlint-disable MD038 -->
!!! failure "Invalid config.yaml"
    ```yaml
    inputs:
      image:
        # Invalid! Thou shalt not mix sequences and mappings in YAML
        - /full/path/to/A/image1.fits
        glob: /full/path/to/B/image*.fits
    ```
<!-- markdownlint-enable MD038 -->

However, an explicit file path is a valid glob expression, so adding explicit paths alongside glob expressions is still possible by simply including the path in a list of glob expressions. For example, the following is valid:

<!-- markdownlint-disable MD038 -->
!!! success "config.yaml"
    ```yaml
    inputs:
      image:
        glob:
        - /full/path/to/A/image1.fits
        - /full/path/to/B/image*.fits
    ```
<!-- markdownlint-enable MD038 -->

In the above example, the final resolved image input list would contain the image `/full/path/to/A/image1.fits`, followed by all files matching `image*.fits` in `/full/path/to/B`.

### Source Monitoring

**`source_monitoring.monitor`**
Boolean. Turns on or off forced extractions for non detections. If set to `True` then `inputs.background` must also be defined. Defaults to `False`.

**`source_monitoring.min_sigma`**
Float. For forced extractions to be performed they must meet a minimum signal-to-noise threshold with respect to the minimum rms value of the respective image. If the proposed forced measurement does not meet the threshold then it is not performed. I.e.

$$
\frac{f_{peak}}{\text{rms}_{image, min}} \geq \small{\text{source_monitoring.min_sigma}}\text{,}
$$

where $f_{peak}$ is the initial detection peak flux measurement of the source in question and $\text{rms}_{image, min}$ is the minimum RMS value of the image where the forced extraction is to take place. Defaults to `3.0`.

**`source_monitoring.edge_buffer_scale`**
Float. Monitor forced extractions are not performed when the location is within 3 beamwidths of the image edge. This parameter scales this distance by the value set, which can help avoid errors when the 3 beamwidth limit is insufficient to avoid extraction failures. Defaults to 1.2.

**`source_monitoring.cluster_threshold`**
Float. A argument directly passed to the [forced photometry package](https://github.com/askap-vast/forced_phot/) used by the pipeline. It defines the multiple of `major_axes` to use for identifying clusters. Defaults to 3.0. To turn off clustering, set it to 0.

**`source_monitoring.allow_nan`**
Boolean. A argument directly passed to the [forced photometry package](https://github.com/askap-vast/forced_phot/) used by the pipeline. It defines whether `NaN` values are allowed to be present in the extraction area in the rms or background maps. `True` would mean that `NaN` values are allowed. Defaults to False.

### Association

!!! tip
    Refer to the [association documentation](../design/association.md) for full details on the association methods.

**`source_association.method`**
String. Select whether to use the `basic`, `advanced` or `deruiter` association method, entered as a string of the method name. Defaults to `"basic"`.

**`source_association.radius`**
Float. The distance limit to use during `basic` and `advanced` association. Unit is arcseconds. Defaults to `10.0`.

**`source_association.deruiter_radius`**
Float. The de Ruiter radius limit to use during `deruiter` association only. The parameter is unitless. Defaults to `5.68`.

**`source_association.deruiter_beamwidth_limit`**

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), with an added `List of PRs` section and links to the relevant PRs on the individual updates. This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/askap-vast/vast-pipeline/compare/v1.2.0...HEAD)

#### Added

- V2: Added functionality to throttle parallel database uploads [#852](https://github.com/askap-vast/vast-pipeline/pull/852)
- V2: Added Dask cluster memory usage logging [#852](https://github.com/askap-vast/vast-pipeline/pull/852)
- V2: Enable specification of Dask worker memory limits [#843](https://github.com/askap-vast/vast-pipeline/pull/843)
- V2: Migrate pipeline to used a Dask.distributed.LocalCluster throughout [#816](https://github.com/askap-vast/vast-pipeline/pull/816)
- V2: Add Dask.distributed support
- V2: Use `django-postgres-copy` for database uploads [#803](https://github.com/askap-vast/vast-pipeline/pull/803)
- Added image.utils.get_fits_header to only fetch header when initialising Image object [#841](https://github.com/askap-vast/vast-pipeline/pull/841)
- Added more detailed warnings about the use of Condon Errors throughout the docs and code [#806](https://github.com/askap-vast/vast-pipeline/pull/806)
- Added processing configuration parameters to the run configuration documentation [#805](https://github.com/askap-vast/vast-pipeline/pull/805)
- Added option to disable forced photometry clustering [#788](https://github.com/askap-vast/vast-pipeline/pull/788)
- Added additional logging to forced extraction step [#788](https://github.com/askap-vast/vast-pipeline/pull/788)

#### Changed

- V2: Revert back to parallel association upload [#852](https://github.com/askap-vast/vast-pipeline/pull/852)
- V2: Change worker throttling to use Dask Semaphore [#852](https://github.com/askap-vast/vast-pipeline/pull/852)
- V2: Compute associations prior to upload, and then upload in serial [#846](https://github.com/askap-vast/vast-pipeline/pull/846)
- V2: Persist srcs_df prior to measurement pairs df output file computation [#846](https://github.com/askap-vast/vast-pipeline/pull/846)
- V2: Change measurement and association DB indices to UUIDs [#844](https://github.com/askap-vast/vast-pipeline/pull/844)
- V2: Limit associations upload to using num_io_workers [#833](https://github.com/askap-vast/vast-pipeline/pull/833)
- V2: Allow user specification of dask dashboard paramters and add some further logging to dask setup [#829](https://github.com/askap-vast/vast-pipeline/pull/829)
- V2: Replace all source, measurement, run etc. IDs with UUID indices. [#803](https://github.com/askap-vast/vast-pipeline/pull/803)
- V2: Updated docs dependencies and fixed outdated information and various warnings [#801](https://github.com/askap-vast/vast-pipeline/pull/801)
- V2: Reorganise new_high_sigma calculation into dedicated function [#714](https://github.com/askap-vast/vast-pipeline/pull/714)
- V2: Update dependencies to fix dependabot warnings and prepare for Dask distributed implementation.
- Speed up final step of basic/advanced association by not resetting index twice (factor of 5-100x faster!) [#861](https://github.com/askap-vast/vast-pipeline/pull/861)
- Speed up wrap handling by not writing to main dataframe [#861](https://github.com/askap-vast/vast-pipeline/pull/861)
- Speed up tmp_srcs_df groupby by not sorting (2x faster) [#861](https://github.com/askap-vast/vast-pipeline/pull/861)
- Speed up various steps by not computing dataframe values multiple times [#861](https://github.com/askap-vast/vast-pipeline/pull/861)
- Converted forced fits and sources removals for re-runs to use raw SQL queries [#840](https://github.com/askap-vast/vast-pipeline/pull/840)
- Updated open_fits to correctly handle NaN-padded compressed images [#837](https://github.com/askap-vast/vast-pipeline/pull/837)
- Upgrade measurements file generation to use dask in order to handle larger runs [#797](https://github.com/askap-vast/vast-pipeline/pull/797)
- Replaced optimise_ints and optimise_floats with single function - optimise_numeric [#797](https://github.com/askap-vast/vast-pipeline/pull/797)
- Upgrade forced_phot dependency to v0.2 and force `use_numba=True` [#788](https://github.com/askap-vast/vast-pipeline/pull/788)
- Remove bad forced photometry fits immediately after calculation, rather than after they've all been compiled into a single dataframe [#788](https://github.com/askap-vast/vast-pipeline/pull/788)
- Optimise associations upload by dropping unnecessary columns prior to a large dataframe merge [#787](https://github.com/askap-vast/vast-pipeline/pull/787)

#### Fixed

- V2: Fixed memory leaks caused by nested delays [#854](https://github.com/askap-vast/vast-pipeline/pull/854)
- V2: Fixed duplicate source ID error by switching from sources_df.repartition() to sources_df.shuffle() in pipeline.finalise.final_operations [#845](https://github.com/askap-vast/vast-pipeline/pull/845)
- V2: Tentative fix for image upload memory leak via garbage collect [#845](https://github.com/askap-vast/vast-pipeline/pull/845)
- V2: Fixed dd.concat memory blow-up in pipeline.forced_extraction.forced_extraction by persisting both dataframes prior [#845](https://github.com/askap-vast/vast-pipeline/pull/845)
- V2: Pair metrics working with V2 dask LocalCluster changes [#817](https://github.com/askap-vast/vast-pipeline/pull/817)
- V2: Fixed missing JS9 overlays on source webpage [#809](https://github.com/askap-vast/vast-pipeline/pull/809)
- V2: Fix bug when deleting source tags from database which are referenced by multiple sources [#803](https://github.com/askap-vast/vast-pipeline/pull/803)
- Fixed incorrect error ellipse calculation [#792](https://github.com/askap-vast/vast-pipeline/pull/792)

webinterface/settings.py
import os
import environ

# Load the Django congig from the .env file
env = environ.Env()
environ.Env.read_env()


# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/2.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = env('SECRET_KEY', cast=str, default='FillMeUPWithSomeComplicatedString')

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = env('DEBUG', cast=bool, default=True)
ALLOWED_HOSTS = env('ALLOWED_HOSTS', cast=list, default=[])
INTERNAL_IPS = [
    '127.0.0.1',
]

SITE_ID = 1

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.humanize',
    'rest_framework',
    'rest_framework_datatables',
    'social_django',
    'crispy_forms',
    'django_q',
    'tagulous',
    # pipeline app and others
    'vast_pipeline',
] + env('EXTRA_APPS', cast=list, default=[])

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'whitenoise.middleware.WhiteNoiseMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'social_django.middleware.SocialAuthExceptionMiddleware',
] + env('EXTRA_MIDDLEWARE', cast=list, default=[])

ROOT_URLCONF = 'webinterface.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [BASE_DIR + '/templates/', ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                'social_django.context_processors.backends',
                'social_django.context_processors.login_redirect',
                'vast_pipeline.context_processors.maintainance_banner',
                'vast_pipeline.context_processors.pipeline_version',
            ],
            'libraries': {
                'unit_tags': 'vast_pipeline.utils.unit_tags'
            }
        },
    },
]

WSGI_APPLICATION = 'webinterface.wsgi.application'

DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'

# Authentication
# Password validation
# https://docs.djangoproject.com/en/3.0/ref/settings/#auth-password-validators

# AUTH_PASSWORD_VALIDATORS = [
#     {
#         'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
#     },
#     {
#         'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
#     },
#     {
#         'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
#     },
#     {
#         'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
#     },
# ]

# docs @ https://python-social-auth.readthedocs.io/en/latest/backends/github.html#github
AUTHENTICATION_BACKENDS = [
    'social_core.backends.github.GithubOrganizationOAuth2',
    'django.contrib.auth.backends.ModelBackend',
]

LOGIN_URL = 'login'
LOGIN_REDIRECT_URL = 'index'
LOGOUT_URL = 'logout'
LOGOUT_REDIRECT_URL = 'login'
LOGIN_ERROR_URL = 'login'

SOCIAL_AUTH_STRATEGY = 'social_django.strategy.DjangoStrategy'
SOCIAL_AUTH_STORAGE = 'social_django.models.DjangoStorage'
SOCIAL_AUTH_POSTGRES_JSONFIELD = True
SOCIAL_AUTH_ADMIN_USER_SEARCH_FIELDS = ['email']
SOCIAL_AUTH_PIPELINE = (
    'social_core.pipeline.social_auth.social_details',
    'social_core.pipeline.social_auth.social_uid',
    'social_core.pipeline.social_auth.auth_allowed',
    'social_core.pipeline.social_auth.social_user',
    'social_core.pipeline.user.get_username',
    'social_core.pipeline.user.create_user',
    'vast_pipeline.utils.auth.create_admin_user',
    'social_core.pipeline.social_auth.associate_user',
    'social_core.pipeline.social_auth.load_extra_data',
    'vast_pipeline.utils.auth.load_github_avatar',
    'social_core.pipeline.user.user_details',
)

SOCIAL_AUTH_GITHUB_ORG_KEY = env('SOCIAL_AUTH_GITHUB_KEY', cast=str, default='')
SOCIAL_AUTH_GITHUB_ORG_SECRET = env('SOCIAL_AUTH_GITHUB_SECRET', cast=str, default='')
SOCIAL_AUTH_GITHUB_ORG_NAME = env('SOCIAL_AUTH_GITHUB_ORG_NAME', cast=str, default='')
SOCIAL_AUTH_GITHUB_ADMIN_TEAM = env('SOCIAL_AUTH_GITHUB_ADMIN_TEAM', cast=str, default='')
SOCIAL_AUTH_REDIRECT_IS_HTTPS = env('SOCIAL_AUTH_REDIRECT_IS_HTTPS', cast=bool, default=False)
SOCIAL_AUTH_GITHUB_ORG_SCOPE = ['read:org', 'user:email']

CRISPY_TEMPLATE_PACK = "bootstrap4"

TNS_API_KEY = env('TNS_API_KEY', default=None)
TNS_USER_AGENT = env('TNS_USER_AGENT', default=None)

# Database
# https://docs.djangoproject.com/en/2.2/ref/settings/#databases

DATABASES = {
    'default': env.db()
}

# Cache (necessary to run pipeline jobs from UI)
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.db.DatabaseCache',
        'LOCATION': 'pipeline_cache_table',
    }
}

# Django Queue Cluster
Q_CLUSTER = {
    'name': 'VastPipeline',
    'workers': 3,
    'timeout': env('Q_CLUSTER_TIMEOUT', cast=int, default=86400),
    'queue_limit': 6,
    'ack_failures': True,
    'bulk': 10,
    'orm': 'default',# same as above in DATABASES but can be changed
    'label': 'Django Q tasks',
    'daemonize_workers': False,
    'recycle': 100,
    'retry': env('Q_CLUSTER_RETRY', cast=int, default=86402),
    'max_attempts': env('Q_CLUSTER_MAX_ATTEMPTS', cast=int, default=1),
}

# REST framework settings
REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.BasicAuthentication',
        'rest_framework.authentication.SessionAuthentication',
    ),
    'DEFAULT_RENDERER_CLASSES': (
        'rest_framework.renderers.JSONRenderer',
        'rest_framework.renderers.BrowsableAPIRenderer',
        'rest_framework_datatables.renderers.DatatablesRenderer',
    ),
    'DEFAULT_FILTER_BACKENDS': (
        'rest_framework_datatables.filters.DatatablesFilterBackend',
    ),
    'DEFAULT_PAGINATION_CLASS': 'rest_framework_datatables.pagination.DatatablesPageNumberPagination',
    'PAGE_SIZE': 100,
}

# Internationalization
# https://docs.djangoproject.com/en/2.2/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/2.2/howto/static-files/
BASE_URL = env('BASE_URL', cast=str, default=None)
STATIC_URL = env('STATIC_URL', cast=str, default='/static/')
if BASE_URL:
    STATIC_URL = '/' + BASE_URL.strip('/') + '/' + STATIC_URL.strip('/') + '/'
STATICFILES_DIRS = env('STATICFILES_DIRS', cast=list, default=[os.path.join(BASE_DIR, 'static')])
STATIC_ROOT = env('STATIC_ROOT', cast=str, default=os.path.join(BASE_DIR, 'staticfiles'))
STATICFILES_STORAGE = 'whitenoise.storage.CompressedManifestStaticFilesStorage'

# Logging
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'verbose': {
        'format': '{asctime} {process:d} {thread:d} {name} {levelname} {message}',
        'style': '{',
        },
        'default': {
        'format': '{asctime} {module} {levelname} {message}',
        'style': '{',
        },
    },
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'class': 'django.utils.log.AdminEmailHandler'
        },
        'console': {
            'class': 'logging.StreamHandler',
            'formatter': 'default',
        },
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
        # root logger
        '': {
            'handlers': ['console'],
            'propagate': True,
            'level': 'INFO',
        },
    }
}

# PRODUCTION SETTINGS
if not DEBUG:
    # ideally you want to check the site rating at https://securityheaders.com/
    # as suggested here https://adamj.eu/tech/2019/04/10/how-to-score-a+-for-security-headers-on-your-django-website/
    # SECURE_SSL_REDIRECT = True # set this to True when your reverse proxy server does not redirect http to https
    SESSION_COOKIE_SECURE = True
    CSRF_COOKIE_SECURE = True
    SECURE_HSTS_SECONDS = 31536000 # see https://docs.djangoproject.com/en/3.1/ref/middleware/#http-strict-transport-security
    SECURE_HSTS_INCLUDE_SUBDOMAINS = True
    SECURE_HSTS_PRELOAD = True
    SECURE_REFERRER_POLICY = 'same-origin' # see https://docs.djangoproject.com/en/3.0/ref/middleware/#referrer-policy
    SECURE_BROWSER_XSS_FILTER = True
    SECURE_CONTENT_TYPE_NOSNIFF = True
    X_FRAME_OPTIONS = 'DENY'
    # from https://ubuntu.com/blog/django-behind-a-proxy-fixing-absolute-urls
    USE_X_FORWARDED_HOST = True
    SECURE_PROXY_SSL_HEADER = ('HTTP_X_FORWARDED_PROTO', 'https')

# PIPELINE settings
# project default folder
PIPELINE_WORKING_DIR = env('PIPELINE_WORKING_DIR', cast=str, default=os.path.join(BASE_DIR, 'pipeline-runs'))
if '/' not in PIPELINE_WORKING_DIR:
    PIPELINE_WORKING_DIR = os.path.join(BASE_DIR, PIPELINE_WORKING_DIR)

# raw image data folder (containing FITS files, selavy, etc)
RAW_IMAGE_DIR = env('RAW_IMAGE_DIR', cast=str, default=os.path.join(BASE_DIR, 'raw-images'))
if '/' not in RAW_IMAGE_DIR:
    RAW_IMAGE_DIR = os.path.join(BASE_DIR, RAW_IMAGE_DIR)

# extra user-supplied data folder
# HOME_DATA_DIR is relative to HOME_DATA_ROOT if HOME_DATA_ROOT is not None
# otherwise, HOME_DATA_DIR is relative to the user's home directory on the deployment machine
HOME_DATA_ROOT = env('HOME_DATA_ROOT', default=None)
HOME_DATA_DIR = env('HOME_DATA_DIR', cast=str, default='vast-pipeline-extra-data')

# allowed source finders
SOURCE_FINDERS = ['selavy']

# default source finder
DEFAULT_SOURCE_FINDER = 'selavy'

# defaults source association methods
DEFAULT_ASSOCIATION_METHODS = ['basic', 'advanced', 'deruiter']

# minimum default accepted error on flux
FLUX_DEFAULT_MIN_ERROR = env('FLUX_DEFAULT_MIN_ERROR', cast=float, default=0.001)

# minimum default accepted error on ra and dec
POS_DEFAULT_MIN_ERROR = env('POS_DEFAULT_MIN_ERROR', cast=float, default=0.01)

# Default pipeline run config values
PIPE_RUN_CONFIG_DEFAULTS = {
    'image_files': [],
    'selavy_files': [],
    'background_files': [],
    'noise_files': [],
    'source_finder': 'selavy',
    'monitor': False,
    'monitor_min_sigma': 3.0,
    'monitor_edge_buffer_scale': 1.2,
    'monitor_cluster_threshold': 3.0,
    'monitor_allow_nan': False,
    'astrometric_uncertainty_ra': 1,
    'astrometric_uncertainty_dec': 1,
    'association_parallel': False,
    'association_epoch_duplicate_radius': 2.5,
    'association_method': 'basic',
    'association_radius': 10.,
    'association_de_ruiter_radius': 5.68,
    'association_beamwidth_limit': 1.5,
    'new_source_min_sigma': 5.0,
    'flux_perc_error': 0,
    'use_condon_errors': True,
    'selavy_local_rms_zero_fill_value': 0.2,
    'create_measurements_arrow_files': False,
    'suppress_astropy_warnings': True,
    'pair_metrics': True,
    'source_aggregate_pair_metrics_min_abs_vs': 4.3,
    'num_workers': 'null',
    'num_workers_io': 5,
    'max_partition_mb': 15
}

# default max concurrent pipeline runs
MAX_PIPELINE_RUNS = env('MAX_PIPELINE_RUNS', cast=int, default=3)

# maximum number of images for non-admin runs
MAX_PIPERUN_IMAGES = env('MAX_PIPERUN_IMAGES', cast=int, default=200)

# maximum number of cutout images to render on the source detail page
MAX_CUTOUT_IMAGES = env('MAX_CUTOUT_IMAGES', cast=int, default=30)

# pipeline maintainance message/banner
PIPELINE_MAINTAINANCE_MESSAGE = env('PIPELINE_MAINTAINANCE_MESSAGE', cast=str, default=None)

# web server eta v plot datashader threshold
ETA_V_DATASHADER_THRESHOLD = env('ETA_V_DATASHADER_THRESHOLD', cast=int, default=20000)

CUTOFF_DAYS = env("CUTOFF_DAYS", cast=int, default=30)
MAX_SIZE_GB = env("MAX_SIZE_GB", cast=float, default=0.01)
MEDIA_ROOT = env("MEDIA_ROOT", cast=str, default=None)
